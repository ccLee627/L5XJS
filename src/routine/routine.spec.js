const Routine = require("./index");
const Element = require("../element");
const Rung = require("../rung");

describe("Routine Class", () => {
    describe("New Instance", () => {
        it("Accepts Proper Input", () => {
            const fn = (name, desc = null) => () => new Routine(name, desc);

            expect(fn()).toThrow();
            expect(fn(12)).toThrow();
            expect(fn("IamRoutine")).not.toThrow();
            expect(fn("IamRoutine", 12)).toThrow();
            expect(fn("IamRoutine", "I am a decription")).not.toThrow();
        });

        it("Initializes with Desired Document Model", () => {
            let routine = new Routine("ARoutine");
            expect(routine).toMatchSnapshot();

            routine = new Routine("ARoutine", "I am a description");
            expect(routine).toMatchSnapshot();
        });
    });

    describe("Methods", () => {
        let rout;

        beforeEach(() => {
            rout = new Routine("IamRoutine");
        });

        test("addRung: Throws on Invalid Input", () => {
            const fn = rung => () => rout.addRung(rung);

            expect(fn(new Rung("XIC(someTag)NOP();"))).not.toThrow();
            expect(fn(new Element({ type: "element", name: "el" }))).toThrow();
        });

        test("addRung: Produces Desired Doc Model", () => {
            const r1 = new Rung("XIC(someTag)NOP();");
            rout.addRung(r1);
            expect(rout).toMatchSnapshot();

            r1.dom.attributes.Number = 1;
            expect(rout).toMatchSnapshot();

            r1.dom.attributes.Number = 0;
            rout.addRung(new Rung("XIC(anotherTag)NOP();"));
            expect(rout).toMatchSnapshot();
        });
    });

    describe("Static Methods", () => {
        test("isRoutine: Returns Appropriate Judgement", () => {
            let routine = new Routine("ARoutine");
            expect(Routine.isRoutine(routine)).toBeTruthy();

            routine = { notARoutine: 12 };
            expect(Routine.isRoutine(routine)).toBeFalsy();

            routine = 12;
            expect(Routine.isRoutine(routine)).toBeFalsy();
        });
    });
});
