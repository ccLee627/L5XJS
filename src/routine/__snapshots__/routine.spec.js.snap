// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routine Class Methods addRung: Produces Desired Doc Model 1`] = `
Routine {
  "_dom": Object {
    "attributes": Object {
      "Name": "IamRoutine",
      "Type": "RLL",
    },
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "attributes": Object {
              "Number": 0,
              "Type": "N",
            },
            "elements": Array [
              Object {
                "elements": Array [
                  Object {
                    "cdata": "XIC(someTag)NOP();",
                    "type": "cdata",
                  },
                ],
                "name": "Text",
                "type": "element",
              },
            ],
            "name": "Rung",
            "type": "element",
          },
        ],
        "name": "RLLContent",
        "type": "element",
      },
    ],
    "name": "Routine",
    "type": "element",
  },
  "class_id": "262fbaac4bd096afccf0fbf1fc161c89",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Routine Class Methods addRung: Produces Desired Doc Model 2`] = `
Routine {
  "_dom": Object {
    "attributes": Object {
      "Name": "IamRoutine",
      "Type": "RLL",
    },
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "attributes": Object {
              "Number": 1,
              "Type": "N",
            },
            "elements": Array [
              Object {
                "elements": Array [
                  Object {
                    "cdata": "XIC(someTag)NOP();",
                    "type": "cdata",
                  },
                ],
                "name": "Text",
                "type": "element",
              },
            ],
            "name": "Rung",
            "type": "element",
          },
        ],
        "name": "RLLContent",
        "type": "element",
      },
    ],
    "name": "Routine",
    "type": "element",
  },
  "class_id": "262fbaac4bd096afccf0fbf1fc161c89",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Routine Class Methods addRung: Produces Desired Doc Model 3`] = `
Routine {
  "_dom": Object {
    "attributes": Object {
      "Name": "IamRoutine",
      "Type": "RLL",
    },
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "attributes": Object {
              "Number": 0,
              "Type": "N",
            },
            "elements": Array [
              Object {
                "elements": Array [
                  Object {
                    "cdata": "XIC(someTag)NOP();",
                    "type": "cdata",
                  },
                ],
                "name": "Text",
                "type": "element",
              },
            ],
            "name": "Rung",
            "type": "element",
          },
          Object {
            "attributes": Object {
              "Number": 1,
              "Type": "N",
            },
            "elements": Array [
              Object {
                "elements": Array [
                  Object {
                    "cdata": "XIC(anotherTag)NOP();",
                    "type": "cdata",
                  },
                ],
                "name": "Text",
                "type": "element",
              },
            ],
            "name": "Rung",
            "type": "element",
          },
        ],
        "name": "RLLContent",
        "type": "element",
      },
    ],
    "name": "Routine",
    "type": "element",
  },
  "class_id": "262fbaac4bd096afccf0fbf1fc161c89",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Routine Class New Instance Initializes with Desired Document Model 1`] = `
Routine {
  "_dom": Object {
    "attributes": Object {
      "Name": "ARoutine",
      "Type": "RLL",
    },
    "elements": Array [
      Object {
        "elements": Array [],
        "name": "RLLContent",
        "type": "element",
      },
    ],
    "name": "Routine",
    "type": "element",
  },
  "class_id": "262fbaac4bd096afccf0fbf1fc161c89",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;

exports[`Routine Class New Instance Initializes with Desired Document Model 2`] = `
Routine {
  "_dom": Object {
    "attributes": Object {
      "Name": "ARoutine",
      "Type": "RLL",
    },
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "cdata": "I am a description",
            "type": "cdata",
          },
        ],
        "name": "Description",
        "type": "element",
      },
      Object {
        "elements": Array [],
        "name": "RLLContent",
        "type": "element",
      },
    ],
    "name": "Routine",
    "type": "element",
  },
  "class_id": "262fbaac4bd096afccf0fbf1fc161c89",
  "elem_id": "33cb4159377b63ce9db11667160e39df",
}
`;
